library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationSource.java

```diff
@@ -4,133 +4,94 @@
+                .checkLocationSettings(
+                        new LocationSettingsRequest.Builder()
+                                .addLocationRequest(locationRequest)
+                                .build()
+                )
+                .addOnCompleteListener(new OnCompleteListener<LocationSettingsResponse>() {
+                    @Override
+                    public void onComplete(@NonNull Task<LocationSettingsResponse> resultTask) {
+                        if (sourceListener != null) sourceListener.onResult(resultTask);
+                    }
+                });
     }
 
-    void startSettingsApiResolutionForResult(Status status, Activity activity) throws SendIntentException {
-        status.startResolutionForResult(activity, RequestCode.SETTINGS_API);
+    void startSettingsApiResolutionForResult(@NonNull ResolvableApiException resolvable, Activity activity) throws SendIntentException {
+        resolvable.startResolutionForResult(activity, RequestCode.SETTINGS_API);
     }
 
     @SuppressWarnings("ResourceType")
     void requestLocationUpdate() {
-        LocationServices.FusedLocationApi.requestLocationUpdates(googleApiClient, locationRequest, this);
+        // This method is suited for the foreground use cases
+        googleApiClient.requestLocationUpdates(locationRequest, this, Looper.myLooper());
     }
 
-    void removeLocationUpdates() {
-        LocationServices.FusedLocationApi.removeLocationUpdates(googleApiClient, this);
+    @NonNull
+    Task<Void> removeLocationUpdates() {
+        return googleApiClient.removeLocationUpdates(this);
     }
 
     @SuppressWarnings("ResourceType")
-    boolean getLocationAvailability() {
-        LocationAvailability locationAvailability =
-              LocationServices.FusedLocationApi.getLocationAvailability(googleApiClient);
-        return locationAvailability != null && locationAvailability.isLocationAvailable();
+    @NonNull
+    Task<LocationAvailability> getLocationAvailability() {
+        return googleApiClient.getLocationAvailability();
     }
 
     @SuppressWarnings("ResourceType")
-    Location getLastLocation() {
-        return LocationServices.FusedLocationApi.getLastLocation(googleApiClient);
-    }
-
-    @Override
-    public void onConnected(@Nullable Bundle bundle) {
-        if (sourceListener != null) sourceListener.onConnected(bundle);
-    }
-
-    @Override
-    public void onConnectionSuspended(int i) {
-        if (sourceListener != null) sourceListener.onConnectionSuspended(i);
+    @NonNull
+    Task<Location> getLastLocation() {
+        return googleApiClient.getLastLocation();
     }
 
     @Override
-    public void onConnectionFailed(@NonNull ConnectionResult connectionResult) {
-        if (sourceListener != null) sourceListener.onConnectionFailed(connectionResult);
-    }
+    public void onLocationResult(LocationResult locationResult) {
+        if (locationResult == null) {
```

Comment:
This logic belongs to provider, please just redirect here.

Reply:
You mean call `sourceListener.onLocationChanged(locationResult)` without loop in source?

Reply:
I mean, create `onLocationResult` method in provider and redirect this one to there, so that logic can be handled there and tested. This class should only be responsible to provide instances and redirect callbacks, so no test is needed. 

```
sourceListener.onLocationResult(locationResult);
```

and then, in that method, have this loop and eventually call `onLocationChanged` from it.